File: dotnet/src/Connectors/Connectors.AI.OpenAI/Tokenizers/Settings/GPT3Settings.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text.Json;
using Microsoft.SemanticKernel.AI;

namespace Microsoft.SemanticKernel.Connectors.AI.OpenAI.Tokenizers.Settings;

internal static class GPT3Settings
{
    /// <summary>Gets the cached encoding table (encoder.json).</summary>
    internal static Dictionary<string, int> Encoder => s_encoder.Value;

    /// <summary>Gets the cached byte pair encoding table (vocab.bpe).</summary>
    internal static Dictionary<(string, string), int> BpeRanks => s_bpeRanks.Value;

    /// <summary>Lazy load the cached encoding table (encoder.json).</summary>
    private static readonly Lazy<Dictionary<string, int>> s_encoder = new(() =>
        JsonSerializer.Deserialize<Dictionary<string, int>>(
            EmbeddedResource.ReadEncodingTable()) ?? throw new AIException(
            AIException.ErrorCodes.InvalidConfiguration,
            "Encoding table deserialization returned NULL"));

    /// <summary>Lazy load the cached byte pair encoding table (vocab.bpe).</summary>
    private static readonly Lazy<Dictionary<(string, string), int>> s_bpeRanks = new(() =>
    {
        string table = EmbeddedResource.ReadBytePairEncodingTable();
        Debug.Assert(table.StartsWith("#version: 0.2", StringComparison.Ordinal));

        // Skip past the header line
        int pos = table.IndexOf('\n') + 1;
        Debug.Assert(pos > 0);

        // For each line, split on the first space and add the pair to the dictionary as a key with the value being the entry number.
        var result = new Dictionary<(string, string), int>();
        int nextPos;
        while ((nextPos = table.IndexOf('\n', pos)) >= 0)
        {
            ReadOnlySpan<char> span = table.AsSpan(pos, nextPos - pos).Trim();
            pos = span.IndexOf(' ');
            if (pos >= 0)
            {
                result.Add((span.Slice(0, pos).ToString(), span.Slice(pos + 1).ToString()), result.Count);
            }

            pos = nextPos + 1;
        }

        return result;
    });
}

File: java/extensions/semantickernel-gpt3-tokenizer/src/main/java/com/microsoft/semantickernel/settings/GPT3Settings.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.settings;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.microsoft.semantickernel.ai.AIException;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import reactor.util.function.Tuple2;
import reactor.util.function.Tuples;

public class GPT3Settings {

    public static final Map<String, Integer> encoder;
    public static final Map<Tuple2<String, String>, Integer> bpeRanks;

    static {
        try {
            encoder = getEncoder();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        try {
            bpeRanks = getBpeRanks();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    // Gets the cached encoding table (encoder.json).
    private static Map<String, Integer> getEncoder() throws IOException {
        try {
            String encodingTableJson = EmbeddedResource.readEncodingTable();

            // Read the JSON file and convert it to a Map
            ObjectMapper objectMapper = new ObjectMapper();
            return objectMapper.readValue(
                    encodingTableJson, new TypeReference<Map<String, Integer>>() {});
        } catch (IOException e) {
            System.out.println(e.getMessage());

            throw new AIException(
                    AIException.ErrorCodes.InvalidConfiguration,
                    "Encoding table deserialization returned NULL");
        }
    }

    // Gets the cached byte pair encoding table (vocab.bpe).

    private static Map<Tuple2<String, String>, Integer> getBpeRanks() throws IOException {
        String table = EmbeddedResource.readBytePairEncodingTable();

        // Skip past the header line
        int pos = table.indexOf('\n') + 1;

        // For each line, split on the first space and add the pair to the map as a key with the
        // value being the entry number.
        Map<Tuple2<String, String>, Integer> result = new HashMap<>();
        int nextPos;
        while ((nextPos = table.indexOf('\n', pos)) >= 0) {
            String line = table.substring(pos, nextPos).trim();
            pos = line.indexOf(' ');
            if (pos >= 0) {
                String first = line.substring(0, pos);
                String second = line.substring(pos + 1);
                result.put(Tuples.of(first, second), result.size());
            }
            pos = nextPos + 1;
        }

        return result;
    }
}
